pilot/pkg/xds/fake.go

```diff
@@ -140,6 +139,27 @@ func NewFakeDiscoveryServer(t test.Failer, opts FakeOptions) *FakeDiscoveryServe
 
 		cg.Store().RegisterEventHandler(schema.Resource().GroupVersionKind(), configHandler)
 	}
+	serviceHandler := func(svc *model.Service, _ model.Event) {
+		pushReq := &model.PushRequest{
+			Full: true,
+			ConfigsUpdated: map[model.ConfigKey]struct{}{{
+				Kind:      gvk.ServiceEntry,
+				Name:      string(svc.Hostname),
+				Namespace: svc.Attributes.Namespace,
+			}: {}},
+			Reason: []model.TriggerReason{model.ServiceUpdate},
+		}
+		s.ConfigUpdate(pushReq)
+	}
+	if err := cg.Registry.AppendServiceHandler(serviceHandler); err != nil {
+		t.Fatalf("append service handler failed: %v", err)
+	}
+	if err := cg.ServiceEntryRegistry.AppendWorkloadHandler(k8s.WorkloadInstanceHandler); err != nil {
+		t.Fatal(err)
+	}
+	if err := k8s.AppendWorkloadHandler(cg.ServiceEntryRegistry.WorkloadInstanceHandler); err != nil {
```

Comment:
You will may need to call `SyncAll` again after this. We call `k8s.Run()` inside of its constructor, so the WorkloadHandler could miss the initial `SyncAll` from inside the k8s controller.

Reply:
This ended up causing a race so I think Ill pass it the handler in the constructor

Reply:
Either that, or fake kube controller should have `Run` called and share the same stop chan as the other controllers.